src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,726 @@
+        pub fn remove(self: *Self, key: Key) void {
+            const set = self.associate(key);
+            const way = self.search(set, key) orelse return;
+
+            self.counts.set(set.offset + way, 0);
+        }
+
+        /// If the key is present in the set, returns the way. Otherwise returns null.
+        inline fn search(self: *Self, set: Set, key: Key) ?usize {
+            const ways = search_tags(set.tags, set.tag);
+
+            var it = BitIterator(Ways){ .bits = ways };
+            while (it.next()) |way| {
+                const count = self.counts.get(set.offset + way);
+                if (count > 0 and equal(key_from_value(set.values[way]), key)) {
+                    return way;
+                }
+            }
+
+            return null;
+        }
+
+        /// Where each set bit represents the index of a way that has the same tag.
+        const Ways = meta.Int(.unsigned, layout.ways);
+
+        inline fn search_tags(tags: *[layout.ways]Tag, tag: Tag) Ways {
+            const x: Vector(layout.ways, Tag) = tags.*;
+            const y: Vector(layout.ways, Tag) = @splat(layout.ways, tag);
+
+            const result: Vector(layout.ways, bool) = x == y;
+            return @ptrCast(*const Ways, &result).*;
+        }
+
+        pub fn put_no_clobber(self: *Self, key: Key) *align(value_alignment) Value {
+            return self.put_no_clobber_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Add a key, evicting an older entry if needed, and return a pointer to the value.
+        /// The key must not already be in the cache.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn put_no_clobber_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (
+                Context,
+                *align(value_alignment) const Value,
+            ) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) *align(value_alignment) Value {
+            const set = self.associate(key);
+
+            if (verify) assert(self.search(set, key) == null);
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            var way = self.clocks.get(clock_index);
+            comptime assert(math.maxInt(@TypeOf(way)) == layout.ways - 1);
+            comptime assert(@as(@TypeOf(way), math.maxInt(@TypeOf(way))) +% 1 == 0);
+
+            const clock_iterations_max = layout.ways * math.maxInt(Count);
+            var safety_count: usize = 1;
+            while (safety_count <= clock_iterations_max + 1) : ({
```

Comment:
```suggestion
            var safety_count: usize = 0;
            const clock_iterations_max = layout.ways * math.maxInt(Count);
            while (safety_count <= clock_iterations_max) : ({
```

Subtracting 1 from both `safety_count` and `clock_iterations_max` should have the same effect?

Reply:
We could then do `else unreachable` on the loop as a simpler way to check that we always break from the loop, and so that we don't need the plus one to cater for the last iteration, which is a tricky case.

Reply:
> Subtracting 1 from both `safety_count` and `clock_iterations_max` should have the same effect?

That is in fact what I originally had and I'd be happy to go back to that. I changed this because you preferred `<=` to `<` in the assert: https://github.com/coilhq/tigerbeetle/pull/85#discussion_r836293072

Reply:
What I'm proposing here is slightly different. Both counts remain counts (instead of index/count as before), and with the `else unreachable` branch there's no longer a need for the `+ 1` to handle the last iteration. i.e. Before there was an assert on the safety counter after the loop. We can drop that with this change.

Reply:
Ah I see, that is indeed cleaner. Thanks!

Reply:
Anytime!